### PR TITLE
Modularized detectors and neuron with enhanced UCF training flow

### PR DESCRIPTION
**Pull request by aliang322**
This PR introduces an enhanced UCF training flow and significant revamping of the `base_miner/` directory, notably changing how detectors are instantiated. The result is a single concise and modular `Miner(BaseMinerNeuron)` in `neurons/miner.py`.

**Key changes**

- Detector and Miner Modularization:
  - New `deepfake_detectors/` with detector implementations that subclass a common abstract base class.
  - New `gating_mechanisms/` with gate implementations that handle content-aware image preprocessing and routing, subclassing a common abstract base class. This is especially useful for multi-agent or mixture-of-expert detectors like `CAMODetector(DeepfakeDetector)` in `deepfake_detectors/camo_detector.py`
  - New `registry.py` which uses a factory method design pattern to instantiate detector and gate subclasses with `Registry`
  - Integration of `Registry` with `miner.py` for dynamic, configurable detector initialization and future extensibility
  - Consolidated multiple `neurons/` miner neurons into a single `neurons/miner.py`.
 
- UCF Training Improvements:
  - Cleaner outputs, now saved to single files and in a single folder per run.
  - New args for hyperparameter tuning.
  - New error messages for diagnosing UCF parameter issues affecting output layer shapes.
  - Config saving during training and loading for detector initialization. Automatic saving for reproducibility and reconstructing correct layer shapes for inference.

- Other:
  - Updated bittensor version `7.3.1` -> `7.4.0`
  - Adhoc unit test scripts for all detectors, gates, and miner.py neuron.
  - Relocated training dataset generation pipelines to the [bitmind-utils repository](https://github.com/BitMind-AI/bitmind-utils/tree/main)
  - Deleted unused UCF legacy scripts.

### CODE REVIEW

File: base_miner/README.md [hunk 1]
```diff
@@ -1,5 +1,42 @@
...
+The `registry.py` file is responsible for managing the creation of detectors and gates using a **Factory Method** design pattern. It auto-registers all `DeepfakeDetector` and `Gate` subclasses from their subfolders to respective `Registry` constants, making it simple to instantiate detectors and gates dynamically based on predefined constants.
+
+- **Factory Pattern**: Ensures a clean, maintainable, and scalable method for creating instances of detectors and gating mechanisms.
+- **Auto-Registration**: Automatically registers all available detector and gate subclasses, enabling a flexible and extensible system.
+
+## Integration with `miner.py`
+
+- **Modular Initialization**: The miner neuron in `bitmind-subnet/neurons/miner.py` leverages the registry system to dynamically initialize the detector used for the forward function, facilitating a highly modular design. The detector module used is determined by neuron config args, defaulting to `"CAMO"`.
+
+## *CAMO: Content-Aware Model Orchestration
```
Comment (benliang99) [inline base_miner/README.md]: Suggestion: I think we can move this CAMO section under another section with header "Base Miner" at the top.

File: base_miner/deepfake_detectors/camo_detector.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            else:
+                raise ValueError(f"Detector {model_name} not found in the registry for {content_type}.")
+
+    def __call__(
+            self, image
+    ) -> float:
+        try:
+            # Determine image content type.
+            content_type, content_data = self.gate(image, use_object_detection=False)
+            print(f"content_type: {content_type}")
```
Comment (benliang99) [inline base_miner/deepfake_detectors/camo_detector.py]: Question: Should these prints be bt.logging or removed?

File: base_miner/gating_mechanisms/gating_mechanism.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
+import torch
+from PIL import Image
+from ultralytics import YOLO
+from base_miner.gating_mechanisms import Gate, FaceGate
+from base_miner import GATE_REGISTRY
+from ultralytics import YOLO
```
Comment (benliang99) [inline base_miner/gating_mechanisms/gating_mechanism.py]: Minor thing: Double import

File: base_miner/gating_mechanisms/gating_mechanism.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        """
+        for gate_name in self.gate_config.keys():
+            try:
+                print(f"Loading {gate_name} gate.")
+                self.gates[gate_name] = self.gate_config[gate_name]()
+            except Exception as e:
+                print(f"Gate {gate_name} not found in the registry.")
+                print(e)
+    
+    def classify_image(self, image, use_object_detection=True):
```
Comment (benliang99) [inline base_miner/gating_mechanisms/gating_mechanism.py]: Question: Should we have `use_object_detection=False` by default? i.e. not using YOLOv8 Reason being that the face detector's `dlib` model already detects faces.

File: base_miner/gating_mechanisms/gating_mechanism.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        Returns:
+            str: Content name, e.g. 'face' if the image contains at least one face, otherwise 'general'.
+            list: List of detected data or None if no content detected
+        """
+        gate_results = {}
+        for gate_name in self.gates.keys():
+            gate_results[gate_name] = {"content": self.gates[gate_name].detect_content_type(image),
+                                        "processed_image": self.gates[gate_name](image)}
+
+        print(f"gate_results: {gate_results}")
```
Comment (benliang99) [inline base_miner/gating_mechanisms/gating_mechanism.py]: Suggestion: delete or change to bt.logging.info

File: base_miner/gating_mechanisms/gating_mechanism.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+        Args:
+            image (PIL.Image): The image to analyze.
+
+        Returns:
+            str: Content name, e.g. 'face' if the image contains at least one face, otherwise 'general'.
+            list: List of detected data or None if no content detected
+        """
+        gate_results = {}
+        for gate_name in self.gates.keys():
```
Comment (benliang99) [inline base_miner/gating_mechanisms/gating_mechanism.py:58]: Clarification: Is this loop redundant given what this function returns?

**Review by benliang99 — COMMENTED** [review]
I have some questions, but overall the structure makes sense to me and is much improved for extensibility. Good work!

File: base_miner/deepfake_detectors/camo_detector.py (hunk 2 above)
Comment (dylanuys) [reply base_miner/deepfake_detectors/camo_detector.py]: we decided to keep the detector classes bittensor agnostic, so we can either remove this or add a `verbose` arg or smth

Comment (aliang322) [reply base_miner/deepfake_detectors/camo_detector.py]: - Thanks, these prints can/will be removed.
- We are no longer using bt.logging in any detector/gate code to ensure a bittensor agnostic design.
  - In a following patch, I will add a logger arg to pass in bt or any custom logging solutions.

File: base_miner/README.md (hunk 1 above)
Comment (aliang322) [reply base_miner/README.md]: Agree

File: base_miner/gating_mechanisms/gating_mechanism.py (hunk 4 above)
Comment (dylanuys) [reply base_miner/gating_mechanisms/gating_mechanism.py]: Agree w Ben, we haven't thoroughly tested YOLOv& yet and should default to dlib

**Review by dylanuys — APPROVED** [review]

File: base_miner/README.md (hunk 1 above)
Comment (aliang322) [reply base_miner/README.md]: Addressed in https://github.com/BitMind-AI/bitmind-subnet/pull/63/commits/bd1c68564b0858d837cbf8e294e0258dbcc9a478

File: base_miner/deepfake_detectors/camo_detector.py (hunk 2 above)
Comment (aliang322) [reply base_miner/deepfake_detectors/camo_detector.py]: Addressed in https://github.com/BitMind-AI/bitmind-subnet/pull/63/commits/3beb37f27f1f49acb58ace27c039f89253547ad7

File: base_miner/gating_mechanisms/gating_mechanism.py (hunk 3 above)
Comment (aliang322) [reply base_miner/gating_mechanisms/gating_mechanism.py]: Addressed as part of https://github.com/BitMind-AI/bitmind-subnet/pull/63/commits/2a240d75ab036e837d7f294b714eeb7c9e86668e

File: base_miner/gating_mechanisms/gating_mechanism.py (hunk 4 above)
Comment (aliang322) [reply base_miner/gating_mechanisms/gating_mechanism.py]: Addressed by parameterizing `object_detection: bool = False` in `GatingMechanism` init and adding a key to `camo.yaml` config file to pass as the arg. Also now avoids loading YOLO unless object detection is required. https://github.com/BitMind-AI/bitmind-subnet/pull/63/commits/2a240d75ab036e837d7f294b714eeb7c9e86668e

File: base_miner/gating_mechanisms/gating_mechanism.py (hunk 5 above)
Comment (aliang322) [reply base_miner/gating_mechanisms/gating_mechanism.py]: Addressed in https://github.com/BitMind-AI/bitmind-subnet/commit/3beb37f27f1f49acb58ace27c039f89253547ad7

**Comment by benliang99** [comment]
LGTM

**Pull request merged**